### PR TITLE
Organization groundwork

### PR DESCRIPTION
**Pull request by pkukkon**
- hoo boy

### CODE REVIEW

File: muikku-elastic-search/src/main/java/fi/otavanopisto/muikku/plugins/search/ElasticSearchProvider.java [hunk 1]
```diff
@@ -211,6 +214,15 @@ public SearchResult searchUsers(String text, String[] textFields, Collection<Env
         query.must(termsQuery("archetype", archetypeNames.toArray(new String[0]))); 
       }
       
+      // TODO: force to have at least one organization?
```
[the remark this answers is not in the document]
Comment (pkukkon) [reply muikku-elastic-search/src/main/java/fi/otavanopisto/muikku/plugins/search/ElasticSearchProvider.java]: Not responsibility of Elastic to determine whether organization(s) are required or not. Needs to be enforced on a case-by-case basis wherever users are searched.

File: muikku-elastic-search/src/main/java/fi/otavanopisto/muikku/plugins/search/ElasticSearchProvider.java [hunk 2]
```diff
@@ -502,6 +514,12 @@ public SearchResult searchWorkspaces(
             .should(boolQuery().mustNot(existsQuery("curriculumIdentifiers")))
             .minimumNumberShouldMatch(1));
       }
+      
+      if (!CollectionUtils.isEmpty(organizationIdentifiers)) {
```
[the remark this answers is not in the document]
Comment (pkukkon) [reply muikku-elastic-search/src/main/java/fi/otavanopisto/muikku/plugins/search/ElasticSearchProvider.java]: Same as above.

File: muikku-elastic-search/src/main/java/fi/otavanopisto/muikku/plugins/search/ElasticSearchProvider.java [hunk 3]
```diff
@@ -593,6 +611,55 @@ public SearchResult searchWorkspaces(String searchTerm, int start, int maxResult
...
+          .setFrom(start)
+          .setSize(maxResults);
+      
+      BoolQueryBuilder boolQuery = boolQuery();
+
+      if (StringUtils.isNotBlank(query)) {
+        boolQuery.must(prefixQuery("name", query));
+      }
+
+      // TODO: force to have at least one organization?
```
[the remark this answers is not in the document]
Comment (pkukkon) [reply muikku-elastic-search/src/main/java/fi/otavanopisto/muikku/plugins/search/ElasticSearchProvider.java]: Same as above

File: muikku-rest/src/main/java/fi/otavanopisto/muikku/rest/model/OrganizationRESTModel.java [hunk 4]
```diff
@@ -1,16 +1,20 @@
 package fi.otavanopisto.muikku.rest.model;
 
-public class UserGroupRESTModel {
+public class OrganizationRESTModel {
```
[the remark this answers is not in the document]
Comment (pkukkon) [reply muikku-rest/src/main/java/fi/otavanopisto/muikku/rest/model/OrganizationRESTModel.java]: User group was modeled both as UserGroupRESTModel (not used anywhere) and UserGroup (used everywhere). Used the former as base for OrganizationRESTModel.

File: muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/search/UserIndexer.java [hunk 5]
```diff
@@ -63,18 +58,14 @@ public void indexUser(String dataSource, String identifier) {
     try {
       User user = userController.findUserByIdentifier(userIdentifier);
       if (user != null) {
-        EnvironmentRoleArchetype archetype = null;
+        // TODO: we have only one role here but a user(entity) can have several roles via several userschooldataidentifiers
```
[the remark this answers is not in the document]
Comment (pkukkon) [reply muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/search/UserIndexer.java]: Agreed, although only applies to staff being students and even then, it is logical that only one of them would be active at any given time.

File: muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/workspace/rest/WorkspaceRESTService.java [hunk 6]
```diff
@@ -446,8 +456,23 @@ else if (userEntity != null) {
         }
       }
       
+      // TODO: Limit to organizations the logged user has access to (how though?)
```
[the remark this answers is not in the document]
Comment (pkukkon) [reply muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/workspace/rest/WorkspaceRESTService.java]: Largely depends on how workspace access will eventually be implemented. Probably somewhat related to sign up rights, e.g. student of organization A may be able to sign up to some workspaces of organization B. Leaving todo as-is to be addressed when new sign up logic is implemented.

File: muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/workspace/rest/WorkspaceRESTService.java [hunk 7]
```diff
@@ -845,6 +870,7 @@ public Response listWorkspaceStudents(@PathParam("WORKSPACEENTITYID") Long works
     // Retrieve users via Elastic
 
     SearchResult searchResult = elasticSearchProvider.searchUsers(
+        organizationEntityController.listUnarchived(),            // organizations
```
[the remark this answers is not in the document]
Comment (pkukkon) [reply muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/workspace/rest/WorkspaceRESTService.java]: Depends on how cross-organization co-operation will work. If workspace has teacher from organization A and students from organizations A and B, it is assumed that the teacher should be able to see everyone in order to be able to guide and evaluate them.

File: muikku-core/src/main/java/fi/otavanopisto/muikku/schooldata/GradingSchoolDataController.java [hunk 8]
```diff
@@ -334,15 +343,21 @@ public WorkspaceAssessmentRequest findLatestAssessmentRequestByStudent(String sc
   }
 
   public WorkspaceAssessment findLatestWorkspaceAssessmentByStudent(String schoolDataSource, String studentIdentifier) {
-    SchoolDataSource dataSource = schoolDataSourceDAO.findByIdentifier(schoolDataSource);
-    GradingSchoolDataBridge schoolDataBridge = getGradingBridge(dataSource);
-    if (schoolDataBridge != null) {
-      return schoolDataBridge.findLatestWorkspaceAssessmentByStudent(studentIdentifier);
+    schoolDataBridgeSessionController.startSystemSession();
```
[the remark this answers is not in the document]
Comment (pkukkon) [reply muikku-core/src/main/java/fi/otavanopisto/muikku/schooldata/GradingSchoolDataController.java]: Used by the timed notification strategy that notifies students who haven't had any evaluation activity in the past three months.

File: muikku-core/src/main/java/fi/otavanopisto/muikku/schooldata/GradingSchoolDataController.java [hunk 9]
```diff
@@ -86,13 +89,19 @@ public WorkspaceAssessment findWorkspaceAssessment(SchoolDataIdentifier workspac
   }
   
   public List<WorkspaceAssessment> listWorkspaceAssessments(SchoolDataSource schoolDataSource, String workspaceIdentifier, String studentIdentifier) {
-    GradingSchoolDataBridge schoolDataBridge = getGradingBridge(schoolDataSource);
-    if (schoolDataBridge != null) {
-      return schoolDataBridge.listWorkspaceAssessments(workspaceIdentifier, studentIdentifier);
-    } else {
-      logger.log(Level.SEVERE, "School Data Bridge could not be found for data source: "  + schoolDataSource.getIdentifier());
+    schoolDataBridgeSessionController.startSystemSession();
```
[the remark this answers is not in the document]
Comment (pkukkon) [reply muikku-core/src/main/java/fi/otavanopisto/muikku/schooldata/GradingSchoolDataController.java]: Most likely guider functionality that assumes that if a teacher can see a student, they should also be able to see all workspaces of that student and the related evaluation info as well.

File: updates/1.089/191029-organization-base.xml [hunk 10]
```diff
@@ -0,0 +1,103 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
[the remark this answers is not in the document]
Comment (pkukkon) [reply updates/1.089/191029-organization-base.xml]: Would be nice housekeeping, sure. Would prefer it as its own issue, though, as the database has lots of other tables too that are no longer used (with or without code).

File: muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/coursepicker/CoursePickerRESTService.java [hunk 11]
```diff
@@ -166,6 +173,30 @@ public Response listCurriculums() {
...
   
+  @GET
+  @Path("/organizations")
+  @RESTPermit (requireLoggedIn = false, handling = Handling.UNSECURED)
+  public Response listOrganizations() {
+    schoolDataBridgeSessionController.startSystemSession();
+    try {
+      List<OrganizationEntity> organizations = organizationEntityController.listUnarchived();
+      
+      // TODO: Limit to organizations the user is part of / has access to
```
[the remark this answers is not in the document]
Comment (pkukkon) [reply muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/coursepicker/CoursePickerRESTService.java]: More related to sign up and cross-organization co-operation logic to be addressed later.

File: muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/coursepicker/CoursePickerRESTService.java [hunk 12]
```diff
@@ -258,8 +290,23 @@ public Response listWorkspaces(
         }
       }
       
+      // TODO: Limit to organizations the logged user has access to (how though?)
```
[the remark this answers is not in the document]
Comment (pkukkon) [reply muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/coursepicker/CoursePickerRESTService.java]: More related to sign up and cross-organization co-operation logic to be addressed later.

File: muikku-atests/src/test/java/fi/otavanopisto/muikku/ui/base/course/CourseTestsBase.java [hunk 13]
```diff
@@ -324,4 +324,59 @@ public void courseProgressWidgetTest() throws Exception {
...
+  @TestEnvironments (
+      browsers = {
+        TestEnvironments.Browser.CHROME,
+        TestEnvironments.Browser.FIREFOX,
+        TestEnvironments.Browser.INTERNET_EXPLORER,
+        TestEnvironments.Browser.EDGE,
+        TestEnvironments.Browser.SAFARI
+      }
+    )
+  public void courseTeacherVacationInfoTest() throws Exception {
```
[the remark this answers is not in the document]
Comment (anttiviljakainen) [reply muikku-atests/src/test/java/fi/otavanopisto/muikku/ui/base/course/CourseTestsBase.java]: @petja-laitila

File: muikku-timed-notifications-plugin/src/main/java/fi/otavanopisto/muikku/plugins/timed/notifications/strategies/AbstractTimedNotificationStrategy.java [hunk 14]
```diff
@@ -13,12 +15,19 @@
...
 import fi.otavanopisto.muikku.jade.JadeController;
+import fi.otavanopisto.muikku.model.users.OrganizationEntity;
 import fi.otavanopisto.muikku.plugins.timed.notifications.TimedNotificationsJadeTemplateLoader;
+import fi.otavanopisto.muikku.schooldata.SchoolDataIdentifier;
+import fi.otavanopisto.muikku.users.OrganizationEntityController;
 
-public abstract class AbstractTimedNotificationStrategy implements TimedNotificationStrategy{
+public abstract class AbstractTimedNotificationStrategy implements TimedNotificationStrategy {
+  
+  private static final String ENABLED_ORGANIZATIONS = System.getProperty("muikku.timednotifications.enabledorganizations");
```
Comment (pkukkon) [inline muikku-timed-notifications-plugin/src/main/java/fi/otavanopisto/muikku/plugins/timed/notifications/strategies/AbstractTimedNotificationStrategy.java]: Requires salt changes? Double-check that notifications continue to work normally even if this hasn't been set (which should be the case, given that it is just a missing search filter?)

Comment (anttiviljakainen) [reply muikku-timed-notifications-plugin/src/main/java/fi/otavanopisto/muikku/plugins/timed/notifications/strategies/AbstractTimedNotificationStrategy.java]: Done

**Pull request merged**